### PR TITLE
Add parity between old and new lambda provider for destinations

### PR DESCRIPTION
**Pull request by dominikschubert**
This will still need a lot of work after the initial release but should for now have parity with the existing implementation in the old non-ASF provider.

### CODE REVIEW

File: localstack/services/awslambda/event_source_listeners/adapters.py [hunk 1]
```diff
@@ -219,7 +219,12 @@ def mapped_callback(ft_result: Future[InvocationResult]) -> None:
 
                 ft.add_done_callback(mapped_callback)
 
-            return 200
+            # they're always synchronous in the ASF provider
+            result = ft.result(timeout=900)
+            if isinstance(result, InvocationError):
```
Comment (dfangl) [inline localstack/services/awslambda/event_source_listeners/adapters.py:224]: Do we have a test for this?

File: tests/integration/awslambda/functions/lambda_destination_handler.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+"""
+
+There are a few well-defined ways to control this function:
+
+1. event echo response (default)
+2. blocking wait mode
+3. fail mode
+4. forwarding to SNS
```
Comment (dfangl) [inline tests/integration/awslambda/functions/lambda_destination_handler.py:8]: I do not think any of this works with this function?

File: tests/integration/cloudformation/resources/test_lambda.py [hunk 3]
```diff
@@ -795,3 +796,127 @@ def wait_logs():
...
+            # ("lambda", "lambda"),
+            # ("eventbridge", "eventbridge")
+        ],
+    )
+    @pytest.mark.aws_validated
+    def test_generic_destination_routing(
+        self, lambda_client, logs_client, deploy_cfn_template, cfn_client, on_success, on_failure
+    ):
+        """
+        This fairly simple template lets us choose between the 4 different destinations for both OnSuccess as well as OnSuccess.
```
Comment (dfangl) [inline tests/integration/cloudformation/resources/test_lambda.py:849]: Both OnSuccess and OnFailure, right?
```suggestion
        This fairly simple template lets us choose between the 4 different destinations for both OnSuccess as well as OnFailure.
```

**Review by dfangl — APPROVED** [review]
Just some minor comments/questions

File: localstack/services/awslambda/event_source_listeners/adapters.py (hunk 1 above)
Comment (dominikschubert) [reply localstack/services/awslambda/event_source_listeners/adapters.py:224]: si

**Pull request merged**